pkg/controller/application/app_controller.go

```diff
@@ -0,0 +1,295 @@
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/appmgmt/interfaces"
+	shimcache "github.com/apache/incubator-yunikorn-k8shim/pkg/cache"
+	applicationclient "github.com/apache/incubator-yunikorn-k8shim/pkg/client/clientset/versioned"
+	appscheme "github.com/apache/incubator-yunikorn-k8shim/pkg/client/clientset/versioned/scheme"
+	appinformers "github.com/apache/incubator-yunikorn-k8shim/pkg/client/informers/externalversions"
+	applister "github.com/apache/incubator-yunikorn-k8shim/pkg/client/listers/yunikorn.apache.org/v1alpha1"
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/common/constants"
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/common/events"
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/log"
+)
+
+type AppManager struct {
+	amProtocol interfaces.ApplicationManagementProtocol
+	controller *Controller
+}
+
+func NewAppManager(amProtocol interfaces.ApplicationManagementProtocol) *AppManager {
+	return &AppManager{
+		amProtocol: amProtocol,
+	}
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) Name() string {
+	return constants.AppManagerHandlerName
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) ServiceInit() error {
+	return nil
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) Start() error {
+	appMgr.controller = NewController(appMgr)
+	appMgr.controller.Run()
+	return nil
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) Stop() {
+	// noop
+}
+
+// handle the updates from the scheduler and sync the status change
+func (appMgr *AppManager) HandleCallbackEvents() func(obj interface{}) {
+	return func(obj interface{}) {
+		if event, ok := obj.(events.ApplicationEvent); ok {
+			if events.AppStateChange == event.GetEvent() {
+				if shimEvent, ok := event.(shimcache.ApplicationStatusChangeEvent); ok {
+					appID := event.GetApplicationID()
+					log.Logger.Info("Status Change callback received",
+						zap.String("app id", appID),
+						zap.String("new status", shimEvent.State))
+					var app = appMgr.amProtocol.GetApplication(appID).(*shimcache.Application)
+					appCRD, err := appMgr.controller.lister.Applications(app.GetTags()[constants.AppTagNamespace]).Get(appID)
+					if err != nil {
+						log.Logger.Warn("Failed to query app CRD for status update",
+							zap.String("Application ID", appID),
+							zap.Error(err))
+						return
+					}
+					crdState := convertShimAppStateToAppCRDState(shimEvent.State)
+					if crdState != "Undefined" {
+						appMgr.updateAppCRDStatus(appCRD, crdState)
+					} else {
+						log.Logger.Error("Invalid status, skip saving it",
+							zap.String("App id", appID))
+					}
+				}
+			}
+		}
+	}
+}
+
+type Controller struct {
+	kubeClient       kubernetes.Interface
+	extensionsClient apiextensionsclient.Interface
+	appClient        applicationclient.Interface
+	informer         cache.SharedIndexInformer
+	lister           applister.ApplicationLister
```

Comment:
is it possible to move most of these things to `apifactory.go`? 
the `APIProvider` is supposed to provide a convenient way to access client libraries.

Reply:
I will remove kubeclient and extensionsClient from here, since I need them only for initialising the other variables. The other ones are all application related and I would keep them here, since we need them to be created only if the CRD feature is turned on.